File: shared/src/iosMain/kotlin/com/architect/kmpappinsights/engine/MSAIApplication.kt
package com.architect.kmpappinsights.engine

import platform.Foundation.NSCoder
import platform.Foundation.decodeObjectForKey
import platform.Foundation.encodeObject

data class MSAIApplication(
    var version: String? = null,
    var build: String? = null,
    var typeId: String? = null
) : MSAIObject(), Serializable {

    fun serializeToDictionary(): MSAIOrderedDictionary {
        val dict = super.serializeToDictionary()
        version?.let { dict["ai.application.ver"] = it }
        build?.let { dict["ai.application.build"] = it }
        typeId?.let { dict["ai.application.typeId"] = it }
        return dict
    }

    companion object {
        fun fromCoder(coder: NSCoder): MSAIApplication {
            val app = MSAIApplication()
            app.version = coder.decodeObjectForKey("self.version") as? String
            app.build = coder.decodeObjectForKey("self.build") as? String
            app.typeId = coder.decodeObjectForKey("self.typeId") as? String
            return app
        }
    }

    fun encodeWithCoder(coder: NSCoder) {
        coder.encodeObject(version, forKey = "self.version")
        coder.encodeObject(build, forKey = "self.build")
        coder.encodeObject(typeId, forKey = "self.typeId")
    }
}

File: native/SwiftInsights/SwiftInsights/MSAIApplication.swift
import Foundation

class MSAIApplication: NSObject, NSCoding {

    var version: String?
    var build: String?
    var typeId: String?

    override init() {
        super.init()
    }

    func serializeToDictionary() -> [String: Any] {
        var dict = [String: Any]()
        if let version = version {
            dict["ai.application.ver"] = version
        }
        if let build = build {
            dict["ai.application.build"] = build
        }
        if let typeId = typeId {
            dict["ai.application.typeId"] = typeId
        }
        return dict
    }

    // NSCoding
    required init?(coder: NSCoder) {
        self.version = coder.decodeObject(forKey: "self.version") as? String
        self.build = coder.decodeObject(forKey: "self.build") as? String
        self.typeId = coder.decodeObject(forKey: "self.typeId") as? String
    }

    func encode(with coder: NSCoder) {
        coder.encode(version, forKey: "self.version")
        coder.encode(build, forKey: "self.build")
        coder.encode(typeId, forKey: "self.typeId")
    }
}
